pkg/manager/member/tikv_upgrader.go

```diff
@@ -14,26 +14,162 @@
 package member
 
 import (
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/pingcap/kvproto/pkg/metapb"
 	"github.com/pingcap/tidb-operator/pkg/apis/pingcap.com/v1alpha1"
+	"github.com/pingcap/tidb-operator/pkg/controller"
 	apps "k8s.io/api/apps/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
 )
 
-type tikvUpgrader struct{}
+const (
+	EvictLeaderBeginTime = "evictLeaderBeginTime"
+	EvictLeaderTimeOut   = 3 * time.Minute
+)
+
+type tikvUpgrader struct {
+	pdControl  controller.PDControlInterface
+	podControl controller.PodControlInterface
+	podLister  corelisters.PodLister
+}
 
 // NewTiKVUpgrader returns a tikv Upgrader
-func NewTiKVUpgrader() Upgrader {
-	return &tikvUpgrader{}
+func NewTiKVUpgrader(pdControl controller.PDControlInterface,
+	podControl controller.PodControlInterface,
+	podLister corelisters.PodLister) Upgrader {
+	return &tikvUpgrader{
+		pdControl:  pdControl,
+		podControl: podControl,
+		podLister:  podLister,
+	}
 }
 
 func (tku *tikvUpgrader) Upgrade(tc *v1alpha1.TidbCluster, oldSet *apps.StatefulSet, newSet *apps.StatefulSet) error {
+	ns := tc.GetNamespace()
+	tcName := tc.GetName()
 	if tc.Status.PD.Phase == v1alpha1.UpgradePhase {
 		_, podSpec, err := GetLastAppliedConfig(oldSet)
 		if err != nil {
 			return err
 		}
 		newSet.Spec.Template.Spec = *podSpec
-	} else {
-		tc.Status.TiKV.Phase = v1alpha1.UpgradePhase
+		return nil
```

Comment:
Why return directly？

Reply:
tikv can't be upgraded when pd is upgrading